searx/search.py

```diff
@@ -530,9 +515,10 @@ def __init__(self, search_query, ordered_plugin_list, request):
         self.request = request
 
     def search(self):
-        if plugins.call(self.ordered_plugin_list, 'pre_search', self.request, self):
-            super(SearchWithPlugins, self).search()
+        plugins = get_plugins()
 
+        plugins.call(self.ordered_plugin_list, 'pre_search', self.request, self)
```

Comment:
Why remove ```if ...: super(SearchWithPlugins, self).search()``` ?

Reply:
Hi @dalf, thanks for taking the time for a review.

1. The method ```isuper(SearchWithPlugins, self).search()``` should be called even if the `self.ordered_plugin_list` is empty .. otherwise the search call is not executed when all plugins are disabled. 

2. It make no sense to call the post hook if the pre hook had an empty list of plugins

I mean, if I'm not wrong, all 3 calls needs to be executed.

````python
        plugins.call(self.ordered_plugin_list, 'pre_search', self.request, self)
        super(SearchWithPlugins, self).search()
        plugins.call(self.ordered_plugin_list, 'post_search', self.request, self)
````
It could be optimized to:

````python
        if self.ordered_plugin_list:
            plugins.call(self.ordered_plugin_list, 'pre_search', self.request, self)

        super(SearchWithPlugins, self).search()

        if self.ordered_plugin_list:
           plugins.call(self.ordered_plugin_list, 'post_search', self.request, self)
````
But this use case is rare, since in most installations at least the _HTTPS rewrite_ plugin is enabled.

What do you think, did I overlooked something?

Reply:
I don't know the right answer right now.

But here, some notes / context:
* This comment is related to https://github.com/asciimoo/searx/pull/2027 (if the PR remains a plugin).
* Except this PR, AFIK, there is no ```pre_search``` plugin.
* This design exists since the [first commit about the plugins](https://github.com/asciimoo/searx/commit/00cc4dcbf44d9ecea89befb08cae4ee5561c4247#diff-02503eba222466e11a3f7ae36debdc5cR328) and [this was kept](https://github.com/asciimoo/searx/commit/67e11c42b973932c8f568d80a0f25bfd7fc150ab#diff-35de57dede17c7660869ae806bbc99b6R367-R368) over time

I think this ```if``` overlaps the answers. For example if you search for ```random string``` you get this:
![image](https://user-images.githubusercontent.com/1594191/85920268-9a232b00-b872-11ea-98e6-9b56dcc67594.png)

Answer were created for privacy reason: see https://github.com/asciimoo/searx/pull/797#issuecomment-269694411

Reply:
@dalf - I haven't seen #733 before and thanks for pointing to [".. to stop if there is an answer already from the answerers"](https://github.com/asciimoo/searx/pull/797#issuecomment-269694411) this was not clear to me.  If I take both into account I come to the conclusion that it was a bug from the first commit.  Maybe I'm wrong, but read by yourself ..

- side node: when I looked at the sources I was a bit confused about the names `answerers` and `answers`, to clarify:

  -  `answerers` is what we talk about, implemented in a python module with the exact same name.
  - `answers` are a part of the result container, containing the _"Answers from WEB engines"_

`answerers` is called in base class's method `Search.search()` and this method returns immediately if there are any _local_ "Answers"

````python

    # do search-request
    def search(self):
        global number_of_searches

        # start time
        start_time = time()

        # answeres ?
        answerers_results = ask(self.search_query)

        if answerers_results:
            for results in answerers_results:
                self.result_container.extend('answer', results)
            return self.result_container

````

What means, the search method of the base class needs to be called, even if there are no plugins activated. Compare old inheritance:


````python
    def search(self):
        plugins = get_plugins()
        if plugins.call(self.ordered_plugin_list, 'pre_search', self.request, self):
            super(SearchWithPlugins, self).search()
        plugins.call(self.ordered_plugin_list, 'post_search', self.request, self)

````

with the patched ..

````python
    def search(self):
        plugins = get_plugins()

        plugins.call(self.ordered_plugin_list, 'pre_search', self.request, self)
        super(SearchWithPlugins, self).search()
        plugins.call(self.ordered_plugin_list, 'post_search', self.request, self)

````

I have to make a test with the old implementation and all plugins disabled ... I guess this use case hasn't ever been tested in the past.

Reply:
Pew .. hard to explain, I will try it in just a view words: 

1. in branch master `plugins.call(self.ordered_plugin_list,...)` returns True when the plugin list is empty / see: https://github.com/asciimoo/searx/blob/master/searx/plugins/__init__.py#L70

2. in branch ep-plugins  `plugins.call(self.ordered_plugin_list,...)` returns an empty list when the plugin list is empty / see: https://github.com/return42/searx/blob/ep-plugins/searx/plugins/__init__.py#L216

Conclusion:

1. the `if` condition in branch master was useless from the beginning since the `plugins.call(..)` always returns a value with a typecast into a _boolean_ `True`.
  
2. the implementation of the method `SearchWithPlugins.search(..)` in branch ep-plugins is functional identical to the one in branch master.

I'm wrong / other meanings?

Reply:
https://github.com/asciimoo/searx/blob/575159b194440052d7b48aa073d7e03c80799c90/searx/plugins/__init__.py#L69-L77

Line 73, ```ret``` is assign to return value of the function (```pre_search``` in case of pre_search plugin)

Reply:
> Line 73, ret is assign to return value of the function (pre_search in case of pre_search plugin).

OK, I see, there is a functional change from:

````python

    def call(  # pylint: disable=missing-function-docstring, no-self-use
            self, ordered_plugin_list, plugin_type, request, *args, **kwargs):
        ret = True
        for plugin in ordered_plugin_list:
            if hasattr(plugin, plugin_type):
                ret = getattr(plugin, plugin_type)(request, *args, **kwargs)
                if not ret:
                    break

        return ret
````

to ..

````python
    def call(self, plugin_list, name, *args, **kwargs):
        ...
        f_list = [ getattr(p, name) for p in plugin_list ]
        for func in f_list:
            ret_val.append(func(*args, **kwargs))
        return ret_val

````

But is this really a problem, do we have any drawback with?  Take into account the _execution model_; that we now have external plugins, should a plugin really cancel the execution of the other plugins .. I don't think so.

> In way, in the current implementation (with the if), answerers are specialized pre_search plugins. The core could even create on the fly a plugin from an answerers.

I didn't grasp all, I have to deal with it more .. I mean: I still do not know if I have damaged anything functionally.

Reply:
> I didn't grasp all, I have to deal with it more .. I mean: I still do not know if I have damaged anything functionally.

Sorry it was more open thought about the searx implementation.

Take make it clear, let's keep the thing as it is in your PR, and if later we need change, it can be another PR.